src/adt/internedstateset.c

```diff
@@ -15,436 +18,1226 @@
+ * or reuse an existing one if available.
+ *
+ * This transfers ownership of r_id and l_id from the caller, so it
+ * takes pointers and writes NO_ID into them. */
+static bool
+intern_treap_node(struct interned_state_set_pool *pool,
+    fsm_state_t state, iss_node_id *p_l_id, iss_node_id *p_r_id,
+	iss_node_id *res_id)
+{
+	/* Transfer ownership. */
+	assert(p_l_id != NULL);
+	iss_node_id l_id = *p_l_id;
+	*p_l_id = NO_ID;
+
+	assert(p_r_id != NULL);
+	iss_node_id r_id = *p_r_id;
+	*p_r_id = NO_ID;
+
+	/* refcount: If iss is saved as a new node, l_id and r_id's
+	 * ownership is transferred to it, otherwise they will be
+	 * released later. */
+	struct iss_node iss = {
+		.id = NO_ID,	/* to be assigned */
+		.refcount = 1,
+		.state = state,
+		.l = l_id,
+		.r = r_id,
+	};
 
 #if LOG_INTERNEDSTATESET > 1
-	fprintf(stderr, "interned_state_set_add: cache is using %d / %d\n",
-	    pool->cache.buckets_used, pool->cache.bucket_count);
+	fprintf(stderr, "%s: interning state: %d, l: %d, r: %d\n",
+	    __func__, state, l_id, r_id);
 #endif
-	if (pool->cache.buckets_used == pool->cache.bucket_count/2) {
-		if (!grow_cache_buckets(pool)) {
-			return NULL;
+
+	const uint64_t hash = hash_iss_node(pool, &iss, true);
+	iss.hash = hash;
+
+	/* Grow or clean up hash table if excessively full. Only grow
+	 * the table if it would still be too full once tombstones are
+	 * removed. The table should always have at least some empty
+	 * buckets or performance will degrade rapidly. */
+	if (pool->cache.used_entries + pool->cache.used_tombstones/2 >= pool->cache.count/2) {
+		if (!grow_or_clean_hash_table(pool)) {
+			return false;
 		}
 	}
 
-	for (i = 0; i < pool->cache.bucket_count; i++) {
-		struct cache_bucket *cb = &pool->cache.buckets[(h + i) & mask];
-		if (cb->iss == NO_ISS) {
-			pool->cache.buckets_used++;
-			cb->hash = h;
-			cb->iss = res;
+	/* Count how many probes were neccesary for hash table
+	 * performance tuning. */
+	size_t probes = 0;
+
+	/* Check hash table for a matching node. If not found,
+	 * insert in an empty bucket. There must be at least one. */
+	const uint64_t mask = pool->cache.count - 1;
+
+#define NO_FIRST_TOMBSTONE_POS ((uint64_t)-1)
+	uint64_t first_tombstone_pos = NO_FIRST_TOMBSTONE_POS;
 
-			add_kid_to_iss_cache(pool, set, state, res);
-			return res;
+	for (uint64_t b_i = 0; b_i < pool->cache.count; b_i++) {
+		const uint64_t b = (hash + b_i) & mask;
+		const iss_node_id b_id = pool->cache.buckets[b].id;
+
+#if LOG_INTERNEDSTATESET > 3
+	fprintf(stderr, "%s: htab[b %lu] -> id %d\n", __func__, b, b_id);
+#endif
+
+		if (b_id == TOMBSTONE_ID) {
+			if (first_tombstone_pos == NO_FIRST_TOMBSTONE_POS) {
+				first_tombstone_pos = b;
+			}
+			continue;
+		} else if (b_id == NO_ID) {
+			/* not found -- reserve node, save, and return */
+			iss_node_id new_id;
+			if (!freelist_pop(pool, &new_id)) {
+				return false;
+			}
+			iss.id = new_id;
+			memcpy(&pool->pool[new_id], &iss, sizeof(iss));
```

Comment:
I think we can just use assignment here, can't we?
```
pool->pool[new_id] = iss;
```

Reply:
Done in 8b5b4447.